client/remoterepo.c

```diff
@@ -106,11 +113,490 @@ _handle_curl_cb(
+    return digest_length * 2 == i;
+}
+
+char 
+TDNF_HexToUint
+    (const char *s)
+{
+    char buf[3];
+    unsigned long val;
+
+    buf[0] = s[0];
+    buf[1] = s[1];
+    buf[2] = 0;
+    val = strtoul(buf, NULL, 16);
+    return (unsigned char)(val&0xff);
+}
+
+uint32_t
+TDNF_New_Metalink_Checksum_From_Hex_Digest
+    (const char *hex_digest,
+    char **ppdigest
+    )
+{
+    uint32_t dwError = 0;
+    char *digest;
+    size_t i;
+    size_t len = 0;
+
+    if(!hex_digest ||
+       !ppdigest)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    len = strlen(hex_digest);
+
+    dwError = TDNFAllocateMemory(1, len/2, (void **)&digest);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    for(i = 0; i < len; i += 2)
+    {
+        digest[i/2] = TDNF_HexToUint(hex_digest + i);
+    }
+    *ppdigest = digest;
+
+cleanup:
+    return dwError;
+
+error:
+    if(digest)
+    {
+        TDNF_SAFE_FREE_MEMORY(digest);
+    }
+    goto cleanup;
+}
+
+uint32_t
+TDNF_New_Metalinkfile
+    (metalink_file_t *fileinfo,
+    metalinkfile ****ml_file
+    )
+{
+    metalinkfile *metalink_file = NULL;
+    uint32_t dwError = 0;
+    dwError = TDNFAllocateMemory(1, sizeof(metalinkfile), (void **)&metalink_file);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    dwError = TDNFAllocateString(fileinfo->name, &(metalink_file->filename));
+    BAIL_ON_TDNF_ERROR(dwError); 
+    
+    metalink_file->digest = NULL;
+    if(fileinfo->checksums)
+    {
+        metalink_digest_alias *digest_alias;
+        for(digest_alias = digest_aliases; digest_alias->alias_name; ++digest_alias)
+        {
+            metalink_checksum_t **metalink_checksum;
+            for(metalink_checksum = fileinfo->checksums; *metalink_checksum; ++metalink_checksum)
+            {
+                if(!strcmp(digest_alias->alias_name, (*metalink_checksum)->type) &&
+                   TDNF_Check_Hex_Digest((*metalink_checksum)->hash, digest_alias->length))
+                {
+                    dwError = TDNFAllocateString((*metalink_checksum)->type, &(metalink_file->type));
+                    BAIL_ON_TDNF_ERROR(dwError);
+                    dwError = TDNF_New_Metalink_Checksum_From_Hex_Digest((*metalink_checksum)->hash,
+                                                                        &(metalink_file->digest));
+                    BAIL_ON_TDNF_ERROR(dwError);
+                    break;
+                }
+            }
+            if(metalink_file->digest)
+            {
+                break;
+            }
+        }
+    }
+    if(fileinfo->resources)
+    {
+        metalink_resource_t **metalink_resource = fileinfo->resources;
+        /* Filter by type if it is non-NULL. In Metalink v3, type
+           includes the type of the resource. In repo, we are only
+           interested in HTTP, HTTPS ,FTP and rsync.
+        */
+        if((*metalink_resource)->type == NULL ||
+           !strcmp((*metalink_resource)->type, "http") ||
```

Comment:
Keep all types (http, https ...) in an array and compare in a loop and ignore case